lib/commons/text/label-virtual.js

```diff
@@ -12,28 +12,37 @@ import findUpVirtual from '../dom/find-up-virtual';
  * @param  {VirtualNode} node The virtual node mapping to the input to test
  * @return {Mixed} String of visible text, or `null` if no label is found
  */
-function labelVirtual(node) {
+function labelVirtual(virtualNode) {
 	var ref, candidate, doc;
 
-	candidate = ariaLabelVirtual(node);
+	candidate = ariaLabelVirtual(virtualNode);
 	if (candidate) {
 		return candidate;
 	}
 
 	// explicit label
-	if (node.actualNode.id) {
-		// TODO: es-module-utils.escapeSelector
-		const id = axe.utils.escapeSelector(node.actualNode.getAttribute('id'));
-		doc = getRootNode(node.actualNode);
+	if (virtualNode.attr('id')) {
+		if (!virtualNode.actualNode) {
+			throw new TypeError(
+				'Cannot resolve explicit label reference for non-DOM nodes'
+			);
+		}
+
+		const id = escapeSelector(virtualNode.attr('id'));
+		doc = getRootNode(virtualNode.actualNode);
 		ref = doc.querySelector('label[for="' + id + '"]');
 		candidate = ref && visible(ref, true);
 		if (candidate) {
 			return candidate;
 		}
 	}
 
-	ref = findUpVirtual(node, 'label');
-	candidate = ref && visible(ref, true);
+	if (!virtualNode.actualNode) {
+		throw new TypeError('Cannot determine if non-DOM node is visible');
+	}
```

Comment:
I'm not sure a generic "always throw" is right. Instead lets have "closest" and "visibleVirtual" decide if throwing is necessary.